onnx/defs/parser.h

```diff
@@ -0,0 +1,381 @@
+    std::string s = literal.value;
+    val = std::stoull(s);
+    return Status::OK();
+  }
+
+  Status Parse(float& val) {
+    Literal literal;
+    CHECK_PARSER_STATUS(Parse(literal));
+    switch (literal.type) {
+      case LiteralType::INT_LITERAL:
+      case LiteralType::FLOAT_LITERAL:
+        val = std::stof(literal.value);
+        break;
+      default:
+        PARSE_ERROR("Unexpected literal type.");
+    }
+    return Status::OK();
+  }
+
+  Status Parse(double& val) {
+    Literal literal;
+    CHECK_PARSER_STATUS(Parse(literal));
+    switch (literal.type) {
+      case LiteralType::INT_LITERAL:
+      case LiteralType::FLOAT_LITERAL:
+        val = std::stod(literal.value);
+        break;
+      default:
+        PARSE_ERROR("Unexpected literal type.");
+    }
+    return Status::OK();
+  }
+
+  // Parse a string-literal enclosed within doube-quotes.
+  Status Parse(std::string& val) {
+    Literal literal;
+    CHECK_PARSER_STATUS(Parse(literal));
+    if (literal.type != LiteralType::STRING_LITERAL)
+      PARSE_ERROR("String value expected, but not found.");
+    val = literal.value;
+    return Status::OK();
+  }
+
+  // Parse an identifier, including keywords. If none found, this will
+  // return an empty-string identifier.
+  Status ParseOptionalIdentifier(std::string& id) {
+    SkipWhiteSpace();
+    auto from = next_;
+    if ((next_ < end_) && (isalpha(*next_) || (*next_ == '_'))) {
+      next_++;
+      while ((next_ < end_) && (isalnum(*next_) || (*next_ == '_')))
+        next_++;
+    }
+    id = std::string(from, next_ - from);
+    return Status::OK();
+  }
+
+  Status ParseIdentifier(std::string& id) {
+    ParseOptionalIdentifier(id);
+    if (id.empty())
+      PARSE_ERROR("Identifier expected but not found.");
+    return Status::OK();
+  }
+
+  Status PeekIdentifier(std::string& id) {
+    SavePos();
+    ParseOptionalIdentifier(id);
+    RestorePos();
+    return Status::OK();
+  }
+
+  Status Parse(KeyWordMap::KeyWord& keyword) {
+    std::string id;
+    CHECK_PARSER_STATUS(ParseIdentifier(id));
+    keyword = KeyWordMap::Lookup(id);
+    return Status::OK();
+  }
+
+ protected:
+  const char* start_;
+  const char* next_;
+  const char* end_;
+  const char* saved_pos_;
+};
+
+class OnnxParser : public ParserBase {
+ public:
+  OnnxParser(const char* cstr) : ParserBase(cstr) {}
+
+  Status Parse(IdList& idlist);
+
+  Status Parse(TensorShapeProto& shape);
+
+  Status Parse(TypeProto& typeProto);
+
+  Status Parse(TensorProto& tensorProto);
+
+  Status Parse(ValueInfoProto& valueinfo);
+
+  Status Parse(ValueInfoList& vilist);
+
+  Status ParseSingleAttributeValue(AttributeProto& attr);
+
+  Status Parse(AttributeProto& attr);
+
+  Status Parse(AttrList& attrlist);
+
+  Status Parse(NodeProto& node);
+
+  Status Parse(NodeList& nodelist);
+
+  Status Parse(GraphProto& graph);
+
+  Status Parse(std::string name, GraphProto& graph);
+
+  Status Parse(ModelProto& model);
```

Comment:
Should these methods be part of the `public` interface?

Reply:
I made some of them private methods, and left some of them as public. In general, I think code constructing parts of a model are common. Not all users are going to construct a complete-model in one go. So, I made what seems a reasonable classification for now. Thanks!